README.rst
=======
disamby
=======


.. image:: https://img.shields.io/pypi/v/disamby.svg
        :target: https://pypi.python.org/pypi/disamby

.. image:: https://img.shields.io/travis/verginer/disamby.svg
        :target: https://travis-ci.org/verginer/disamby

.. image:: https://readthedocs.org/projects/disamby/badge/?version=latest
        :target: https://disamby.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status

.. image:: https://pyup.io/repos/github/verginer/disamby/shield.svg
     :target: https://pyup.io/repos/github/verginer/disamby/
     :alt: Updates

* Free software: MIT license
* Documentation: https://disamby.readthedocs.io.

``disamby`` is a python package designed to carry out entity disambiguation based on fuzzy
string matching.

It works best for entities which if the same have very similar strings.
Examples of situation where this disambiguation algorithm works fairly well is with
company names and addresses which have typos, alternative spellings or composite names.
Other use-cases include identifying people in a database where the name might be misspelled.

The algorithm works by exploiting how informative a given word/token is, based on the
observed frequencies in the whole corpus of strings. For example the word 'inc' in the
case of firm names is not very informative, however "Solomon" is, since the former appears
repeatedly whereas the second rarely.

With these frequencies the algorithms computes for a given pair of instances how similar
they are, and if they are above an arbitrary threshold they are connected in an
"alias graph" (i.e. a directed network where an entity is connected to an other
if it is similar enough). After all records have been connected in this way disamby
returns sets of entities, which are strongly connected [2]_ . Strongly connected means
in this case that there exists a path from all nodes to all nodes within the component.


Example
-------

To use disamby in a project::

    import pandas as pd
    import disamby.preprocessors as pre
    form disamby import Disamby

    # create a dataframe with the fields you intend to match on as columns
    df = pd.DataFrame({
        'name':     ['Luca Georger',        'Luca Geroger',         'Adrian Sulzer'],
        'address':  ['Mira, 34, Augsburg',  'Miri, 34, Augsburg',   'Milano, 34']},
        index=      ['L1',                  'L2',                   'O1']
    )

    # define the pipeline to process the strings, note that the last step must return
    # a tuple of strings
    pipeline = [
        pre.normalize_whitespace,
        pre.remove_punctuation,
        lambda x: pre.trigram(x) + pre.split_words(x)  # any python function is allowed
    ]

    # instantiate the disamby object, it applies the given pre-processing pipeline and
    # computes their frequency.
    dis = Disamby(df, pipeline)

    # let disamby compute disambiguated sets. Node that a threshold must be given or it
    # defaults to 0.
    dis.disambiguated_sets(threshold=0.5)
    [{'L2', 'L1'}, {'O1'}]  # output

    # To check if the sets are accurate you can get the rows from the
    # pandas dataframe like so:
    df.loc[['L2', 'L1']]



Installation
------------

To install disamby, run this command in your terminal:

.. code-block:: console

    $ pip install disamby

This is the preferred method to install disamby, as it will always install the most recent stable release.
If you don't have `pip`_ installed, this `Python installation guide`_ can guide
you through the process.

.. _pip: https://pip.pypa.io
.. _Python installation guide: http://docs.python-guide.org/en/latest/starting/installation/

You can also install it from source as follows
The sources for disamby can be downloaded from the `Github repo`_.
You can either clone the public repository:

.. code-block:: console

    $ git clone git://github.com/verginer/disamby

Or download the `tarball`_:

.. code-block:: console

    $ curl  -OL https://github.com/verginer/disamby/tarball/master

Once you have a copy of the source, you can install it with:

.. code-block:: console

    $ python setup.py install


.. _Github repo: https://github.com/verginer/disamby
.. _tarball: https://github.com/verginer/disamby/tarball/master


Credits
---------
I got the inspiration for this package from the seminar "The SearchEngine - A Tool for
Matching by Fuzzy Criteria" by Thorsten Doherr at the CISS [1]_ Summer School 2017

.. [1] http://www.euro-ciss.eu/ciss/home.html
.. [2] https://en.wikipedia.org/wiki/Strongly_connected_component


HISTORY.rst
=======
History
=======


0.2.3 (2017-07-01)
------------------

* Fixes formatting breaking pypi display


0.2.2 (2017-06-30)
------------------

* working release with minimal documentation
* works with multiple field matching
* carries out all steps autonomously from string pre-processing to
  identifying the strongly connected components


0.1.0 (2017-06-24)
------------------

* First release on PyPI.


tests/__init__.py
# -*- coding: utf-8 -*-

"""Unit test package for disamby."""


tests/conftest.py
import os
import pytest
import pandas as pd
import numpy as np
from faker import Faker


TEST_DIR = os.path.dirname(__file__)
DATA_DIR = os.path.join(TEST_DIR, 'data')


@pytest.fixture
def fake_names():
    """function creating fake names of size n"""
    def f_names(n):
        fake = Faker()
        fake.seed(n)
        names = [fake.address() for _ in range(n)]
        return names
    return f_names


@pytest.fixture
def company_df(fake_names):
    """test company details"""
    def df_of_size(size):
        np.random.seed(size)
        fk_names = fake_names(size)
        assignee_path = os.path.join(DATA_DIR, 'potential_assginees_names.csv')
        sample = pd.read_csv(assignee_path, index_col='inv_id').sample(size)
        sample['address'] = fk_names
        del sample['patents']
        return sample
    return df_of_size


tests/test_cli.py
# from click.testing import CliRunner
# from disamby import cli
#
#
# def test_command_line_interface():
#     """Test the CLI."""
#     # TODO: use temporary directory to do actual testing
#     runner = CliRunner()
#     result = runner.invoke(cli.main, [])
#     assert result.exit_code == 2  # missing parameters
#     help_result = runner.invoke(cli.main, ['--help'])
#     assert help_result.exit_code == 0
#     assert 'Show this message and exit.' in help_result.output


disamby/cli.py
# -*- coding: utf-8 -*-

"""Console script for disamby.

With the cli it is possible to carry out the disambiguation on the command line.
The script returns a json file with a name taken from the first field as the key and
a list of indices belonging to the same disambiguated cluster.

.. code-block:: python

    {'International Business Machines Inc': [1, 2, 5, 34, 90],
    'Samsung': [4, 123],
    '...': [...],
     ...}

To carry out the disambiguation you will need a csv file with proper headings
and optionally an index column. If not index column is specified then the position in the
csv is used in the json file to identify its members.

Only a basic processing pipeline is possible here, see the `--help` of the script
to know which are available.

Example usage:

.. code-block:: bash

    $ disamby --col-headers name,address \\
                --index inv_id \\
                --threshold .7 \\
                --prep APNX \\
                input.csv output.json

"""

import click
from disamby import Disamby
import disamby.preprocessors as pre
import pandas as pd
import json

# TODO: Make tests for cli


@click.command(help="""
This command allows to disambiguate a csv file using `disamby`
example usage:\n
$ disamby -c name -i inv_id -t .7 -p APNX name.csv dis_name.json
$ disamby -c name -i inv_id -t .7 -p APNX name.csv dis_name.json

The prep codes are:\n
'A': pre.compact_abbreviations\n
'P': pre.remove_punctuation\n
'W': pre.normalize_whitespace\n
'3': lambda x: pre.ngram(x, 3)\n
'4': lambda x: pre.ngram(x, 4)\n
'5': lambda x: pre.ngram(x, 5)\n
'S': pre.split_words\n
'X': lambda x: pre.ngram(x[:33], 4) + pre.split_words(x)
""")
@click.argument('data', type=click.Path(dir_okay=False, exists=True))
@click.argument('output', type=click.Path(dir_okay=False))
@click.option('-c', '--col-headers', type=click.STRING,
              help='name of the columns containing the strings to match if more then '
                   'one then separate with comma (no spaces)')
@click.option('-t', '--threshold', type=click.FLOAT, default=0.7,
              help='Minimum score to be counted')
@click.option('-i', '--index', type=click.STRING, help='name of index column')
@click.option('-p', '--prep', type=click.STRING, default='APWX',
              help='pre-processing instructions, if left blank will default to a standard'
                   ' prep')
def main(data, output, index, col_headers, prep, threshold):
    """Console script for disamby."""
    names_df = pd.read_csv(data, index_col=index)
    prep_dict = {
        'A': pre.compact_abbreviations,
        'P': pre.remove_punctuation,
        'W': pre.normalize_whitespace,
        '3': lambda x: pre.ngram(x, 3),
        '4': lambda x: pre.ngram(x, 4),
        '5': lambda x: pre.ngram(x, 5),
        'S': pre.split_words,
        'X': lambda x: pre.ngram(x[:33], 4) + pre.split_words(x)
    }
    columns = col_headers.split(',')
    pipeline = [prep_dict[action] for action in list(prep)]
    dis = Disamby(data=names_df[columns], preprocessors=pipeline)
    components = dis.disambiguated_sets(threshold, smoother='offset', offset=100)

    comp_to_id = dict()
    for comp in components:
        members = list(comp)
        representative = members[0]
        name = names_df.loc[representative, columns[0]]
        comp_to_id[name] = members

    with open(output, 'w') as f:
        json.dump(comp_to_id, f)


if __name__ == '__main__':
    main()




disamby/disamby.py
# -*- coding: utf-8 -*-

"""Main module."""
from collections import Counter
from typing import Union
from pandas import DataFrame, Series
from math import log
from tqdm import tqdm
from networkx import DiGraph
from collections import namedtuple

ScoredElement = namedtuple('ScoredElement', ['index', 'score'])
PandasObj = Union[DataFrame, Series]


class Disamby(object):
    """
    Class for disambiguation fitting, scoring and ranking of potential matches

    A `Disamby` instance stores the pre-processing pipeline applied to the
    strings for a given field as well as the the computed frequencies from the
    entire corpus of strings to match against.
    `Disamby` can be instantiated either with not arguments, with a list of
    strings, pandas.Series or pandas.DataFrame. This triggers the immediate
    call to the `fit` method, whose doc explains the parameters.

    Examples
    --------
    >>> import pandas as pd
    >>> import disamby.preprocessors as pre
    >>> df = pd.DataFrame(
    ... {'a': ['Luca Georger', 'Luca Geroger', 'Adrian Sulzer'],
    ... 'b': ['Mira, 34, Augsburg', 'Miri, 34, Augsburg', 'Milano, 34']
    ... }, index=['L1', 'L2', 'O1']
    ... )
    >>> pipeline = [
    ...     pre.normalize_whitespace,
    ...     pre.remove_punctuation,
    ...     pre.trigram
    ... ]
    >>> dis = Disamby(df, pipeline)
    >>> dis.disambiguated_sets(threshold=0.5, verbose=False)
    [{'L2', 'L1'}, {'O1'}]
    """

    def __init__(self, data: PandasObj=None, preprocessors: list=None, field: str=None):
        self.field_freq = dict()
        self.preprocessors = dict()
        self.records = dict()
        self._processed_token_cache = dict()
        self._most_common = dict()
        self._token_to_instance = dict()
        if data is not None:
            if preprocessors is None:
                raise ValueError("Preprocessor not provided")
            self.fit(data, preprocessors, field)

    def fit(self, data: PandasObj, preprocessors: list, field: str=None):
        """
        Computes the frequencies of the terms by field.

        Parameters
        ----------
        data : pandas.DataFrame, pandas.Series or list of strings
            list of strings or pandas.DataFrame
            if dataframe is given then the field defaults to the column name
        preprocessors : list
            list of functions to apply in that order
            note the first function must accept a string, the other functions
            must be such that a pipeline is possible the result is a tuple of
            strings.
        field : str
            string identifying which field this data belongs to

        Examples
        --------

        >>> import pandas as pd
        >>> from disamby.preprocessors import split_words
        >>> df = pd.DataFrame(
        ... {'a': ['Luca Georger', 'Luke Geroge', 'Adrian Sulzer'],
        ... 'b': ['Mira, 34, Augsburg', 'Miri, 32', 'Milano, 34']
        ... })
        >>> dis = Disamby()
        >>> prep = [split_words]
        >>> dis.fit(df, prep)
        """
        try:
            columns = data.columns
            for col in columns:
                self._fit_field(data[col], preprocessors=preprocessors)
        except AttributeError:
            self._fit_field(data, preprocessors=preprocessors, field=field)

    def find(self, idx, threshold=0.0, weights: dict=None, **kwargs) -> list:
        """
        returns the list of scored instances which have a score above the
        threshold. Note that strings which do not share any token are omitted
        since their score is 0 by default.

        Parameters
        ----------
        idx
            index of the record to find
        threshold
        weights : dict

        Returns
        -------

        """
        fields = self.fields

        scored_candidates = dict()
        if weights is None:
            weights = {f: 1 / len(fields) for f in fields}

        for field in fields:
            own_term = self.records[field][idx]
            own_tokens = self._processed_token_cache[field][own_term]
            potential_candidates = set()

            for token in own_tokens:
                potential_candidates |= self._token_to_instance[field][token]

            for candidate in potential_candidates:
                score = self.score(own_term, candidate[1], field, **kwargs)
                candidate_idx = candidate[0]
                if candidate_idx not in scored_candidates:
                    scored_candidates[candidate_idx] = {field: score * weights[field]}
                else:
                    scored_candidates[candidate_idx][field] = score * weights[field]

        # compute weighted score
        final_candidates = []
        for idx, scores in scored_candidates.items():
            total_score = sum(scores.values())
            if total_score >= threshold:
                final_candidates.append(ScoredElement(idx, total_score))

        return final_candidates

    def _fit_field(self, data: PandasObj, preprocessors: list=None, field: str=None):
        if field not in self.preprocessors:
            ValueError('preprocessors have already been defined, '
                       'cannot redefine. This would render the lookup '
                       'inconsistent')

        if field is None:
            try:
                field = data.name
            except AttributeError:  # was not a pandas.Series
                raise KeyError("The provided data are not a pandas Series, "
                               "if the data is a list you need to provide the"
                               "`field` argument.")

        self.preprocessors[field] = preprocessors
        self._processed_token_cache[field] = dict()
        self._token_to_instance[field] = dict()
        self.records[field] = dict()
        counter = Counter()

        for i, name in data.items():
            norm_tokens = self.pre_process(name, preprocessors)
            self._processed_token_cache[field][name] = norm_tokens
            counter.update(norm_tokens)
            self.records[field][i] = name
            for token in norm_tokens:
                if token in self._token_to_instance[field]:
                    self._token_to_instance[field][token] |= {(i, name)}
                else:
                    self._token_to_instance[field][token] = {(i, name)}
        self._most_common[field] = counter.most_common(1)[0][1]
        self.field_freq[field] = counter

    def score(self, term: str, other_term: str, field: str,
              smoother=None, offset=0) -> float:
        """
        Computes the score between the two strings using the frequency data

        Parameters
        ----------
        term : str
            term to search for
        other_term : str
            the other term to compare too
        field : str
            the name of the column to which this term belongs
        smoother : str (optional)
            one of {None, 'offset', 'log'}
        offset : int
            offset to add to count only needed for smoothers 'log' and 'offset'

        Returns
        -------
        float

        Notes
        -----
        The score is not commutative (i.e. score(A,B) != score(B,A))
        """

        try:
            own_parts = self._processed_token_cache[field][term]
            other_parts = self._processed_token_cache[field][other_term]
        except KeyError:
            own_parts = self.pre_process(term, self.preprocessors[field])
            other_parts = self.pre_process(other_term, self.preprocessors[field])

        # get list of potential scores
        weights = self.id_potential(own_parts, field, smoother, offset)
        score = sum(weights.get(tok, 0) for tok in other_parts)
        return score

    def id_potential(self, term: Union[tuple, str], field: str,
                     smoother: str = None, offset=0) -> dict:
        """
        Computes the weights of the words based on the observed frequency
        and normalized.

        Parameters
        ----------
        term : str, tuple
            term to look for or a tuple of proper tokens
        field : str
            field the word falls under
        smoother : str (optional)
            one of {None, 'offset', 'log'}
        offset : int
            offset to add to count only needed for smoothers 'log' and 'offset'

        Returns
        -------
        float
        """
        if isinstance(term, str):
            words = self.pre_process(term, self.preprocessors[field])
        else:
            words = term

        smoothers = {
            None: self._smooth_none,
            'offset': self._smooth_offset,
            'log': self._smooth_log
        }
        if smoother not in smoothers:
            raise KeyError(
                'Chosen smother `{}` is not one of {}'.format(
                    smoother, smoothers.keys())
            )
        counter = self.field_freq[field]

        s_fun = smoothers[smoother]
        max_occ = self._most_common[field]
        id_potentials = {
            word: s_fun(counter[word], offset, max_occ)
            for word in words
        }

        total_weight = sum(id_potentials.values())
        return {w: idp / total_weight for w, idp in id_potentials.items()}

    def alias_graph(self, threshold=0.7, verbose=True, weights=None, **kwargs) -> DiGraph:
        """
        This function creates the directed network connecting an instance to an other
        through a directed edge if the the target instance has a similarity score above
        the threshold.

        Parameters
        ----------
        weights
        threshold : float
            between 0 and 1
        verbose : whether to show the progressbar
        kwargs :
            arguments to pass to the score function (i.e. offset, smoother)

        Returns
        -------
        DiGraph

        """
        if not verbose:
            t = lambda x: x
        else:
            t = tqdm

        edges = []
        fields = self.records.keys()
        a_field = list(fields)[0]
        for idx in t(self.records[a_field]):
            targets = self.find(idx, threshold=threshold, weights=weights, **kwargs)
            new_edges = [(idx, x.index, {'score': x.score}) for x in targets]
            edges.extend(new_edges)

        alias_graph = DiGraph()
        alias_graph.add_edges_from(edges)
        return alias_graph

    def disambiguated_sets(self, threshold=0.7, verbose=True, weights=None, **kwargs):
        from networkx import strongly_connected_components
        alias_graph = self.alias_graph(threshold, verbose, weights, **kwargs)
        strong = strongly_connected_components(alias_graph)
        return list(strong)

    @staticmethod
    def pre_process(base_name, functions: list):
        """apply every function consecutively to base_name"""
        norm_name = base_name
        for f in functions:
            norm_name = f(norm_name)
        return set(norm_name)

    @property
    def fields(self):
        return self.field_freq.keys()

    @staticmethod
    def _smooth_none(occurrences, *args):
        return 1 / max(occurrences, 1)

    @staticmethod
    def _smooth_offset(occurrences, offset, *args):
        return 1 / max(occurrences + offset, 1)

    @staticmethod
    def _smooth_log(occ, offset, max_occ):
        max_offset = max(max_occ + offset + 1, 1)
        word_offset = max(occ + offset, 1)
        return log(max_offset / word_offset)




disamby/core.py
# -*- coding: utf-8 -*-

"""Main module."""
from collections import Counter
from typing import Union
from pandas import DataFrame, Series
from math import log
from tqdm import tqdm
from networkx import DiGraph
from collections import namedtuple

ScoredElement = namedtuple('ScoredElement', ['index', 'score'])
PandasObj = Union[DataFrame, Series]


class Disamby(object):
    """
    Class for disambiguation fitting, scoring and ranking of potential matches

    A `Disamby` instance stores the pre-processing pipeline applied to the
    strings for a given field as well as the the computed frequencies from the
    entire corpus of strings to match against.
    `Disamby` can be instantiated either with not arguments, with a list of
    strings, pandas.Series or pandas.DataFrame. This triggers the immediate
    call to the `fit` method, whose doc explains the parameters.

    Examples
    --------
    >>> import pandas as pd
    >>> import disamby.preprocessors as pre
    >>> df = pd.DataFrame(
    ... {'a': ['Luca Georger', 'Luca Geroger', 'Adrian Sulzer'],
    ... 'b': ['Mira, 34, Augsburg', 'Miri, 34, Augsburg', 'Milano, 34']
    ... }, index=['L1', 'L2', 'O1']
    ... )
    >>> pipeline = [
    ...     pre.normalize_whitespace,
    ...     pre.remove_punctuation,
    ...     pre.trigram
    ... ]
    >>> dis = Disamby(df, pipeline)
    >>> dis.disambiguated_sets(threshold=0.5, verbose=False)
    [{'L2', 'L1'}, {'O1'}]
    """

    def __init__(self, data: PandasObj=None, preprocessors: list=None, field: str=None):
        self.field_freq = dict()
        self.preprocessors = dict()
        self.records = dict()
        self._processed_token_cache = dict()
        self._most_common = dict()
        self._token_to_instance = dict()
        if data is not None:
            if preprocessors is None:
                raise ValueError("Preprocessor not provided")
            self.fit(data, preprocessors, field)

    def fit(self, data: PandasObj, preprocessors: list, field: str=None):
        """
        Computes the frequencies of the terms by field.

        Parameters
        ----------
        data : pandas.DataFrame, pandas.Series or list of strings
            list of strings or pandas.DataFrame
            if dataframe is given then the field defaults to the column name
        preprocessors : list
            list of functions to apply in that order
            note the first function must accept a string, the other functions
            must be such that a pipeline is possible the result is a tuple of
            strings.
        field : str
            string identifying which field this data belongs to

        Examples
        --------

        >>> import pandas as pd
        >>> from disamby.preprocessors import split_words
        >>> df = pd.DataFrame(
        ... {'a': ['Luca Georger', 'Luke Geroge', 'Adrian Sulzer'],
        ... 'b': ['Mira, 34, Augsburg', 'Miri, 32', 'Milano, 34']
        ... })
        >>> dis = Disamby()
        >>> prep = [split_words]
        >>> dis.fit(df, prep)
        """
        try:
            columns = data.columns
            unique_columns = len(set(columns)) == len(columns)
            if not unique_columns:
                raise KeyError("Some columns have identical names")
        except AttributeError:
            self._fit_field(data, preprocessors=preprocessors, field=field)
            return None

        for col in columns:
            self._fit_field(data[col], preprocessors=preprocessors)

    def find(self, idx, threshold=0.0, weights: dict=None, **kwargs) -> list:
        """
        returns the list of scored instances which have a score above the
        threshold. Note that strings which do not share any token are omitted
        since their score is 0 by default.

        Parameters
        ----------
        idx
            index of the record to find
        threshold
        weights : dict

        Returns
        -------

        """
        fields = self.fields

        scored_candidates = dict()
        if weights is None:
            weights = {f: 1 / len(fields) for f in fields}

        for field in fields:
            own_term = self.records[field][idx]
            own_tokens = self._processed_token_cache[field][own_term]
            potential_candidates = set()

            for token in own_tokens:
                potential_candidates |= self._token_to_instance[field][token]

            for candidate in potential_candidates:
                score = self.score(own_term, candidate[1], field, **kwargs)
                candidate_idx = candidate[0]
                if candidate_idx not in scored_candidates:
                    scored_candidates[candidate_idx] = {field: score * weights[field]}
                else:
                    scored_candidates[candidate_idx][field] = score * weights[field]

        # compute weighted score
        final_candidates = []
        for idx, scores in scored_candidates.items():
            total_score = sum(scores.values())
            if total_score >= threshold:
                final_candidates.append(ScoredElement(idx, total_score))

        return final_candidates

    def _fit_field(self, data: PandasObj, preprocessors: list=None, field: str=None):
        if field not in self.preprocessors:
            ValueError('preprocessors have already been defined, '
                       'cannot redefine. This would render the lookup '
                       'inconsistent')

        if field is None:
            try:
                field = data.name
            except AttributeError:  # was not a pandas.Series
                raise KeyError("The provided data are not a pandas Series, "
                               "if the data is a list you need to provide the"
                               "`field` argument.")

        self.preprocessors[field] = preprocessors
        self._processed_token_cache[field] = dict()
        self._token_to_instance[field] = dict()
        self.records[field] = dict()
        counter = Counter()

        sample_item = data.iloc[0]
        if not isinstance(sample_item, str):
            raise ValueError('type of field/column "%s" is not `str`' % field)

        for i, name in data.items():
            norm_tokens = self.pre_process(name, preprocessors)
            self._processed_token_cache[field][name] = norm_tokens
            counter.update(norm_tokens)
            self.records[field][i] = name
            for token in norm_tokens:
                if token in self._token_to_instance[field]:
                    self._token_to_instance[field][token] |= {(i, name)}
                else:
                    self._token_to_instance[field][token] = {(i, name)}
        self._most_common[field] = counter.most_common(1)[0][1]
        self.field_freq[field] = counter

    def score(self, term: str, other_term: str, field: str,
              smoother=None, offset=0) -> float:
        """
        Computes the score between the two strings using the frequency data

        Parameters
        ----------
        term : str
            term to search for
        other_term : str
            the other term to compare too
        field : str
            the name of the column to which this term belongs
        smoother : str (optional)
            one of {None, 'offset', 'log'}
        offset : int
            offset to add to count only needed for smoothers 'log' and 'offset'

        Returns
        -------
        float

        Notes
        -----
        The score is not commutative (i.e. score(A,B) != score(B,A))
        """

        try:
            own_parts = self._processed_token_cache[field][term]
            other_parts = self._processed_token_cache[field][other_term]
        except KeyError:
            own_parts = self.pre_process(term, self.preprocessors[field])
            other_parts = self.pre_process(other_term, self.preprocessors[field])

        # get list of potential scores
        weights = self.id_potential(own_parts, field, smoother, offset)
        score = sum(weights.get(tok, 0) for tok in other_parts)
        return score

    def id_potential(self, term: Union[tuple, str], field: str,
                     smoother: str = None, offset=0) -> dict:
        """
        Computes the weights of the words based on the observed frequency
        and normalized.

        Parameters
        ----------
        term : str, tuple
            term to look for or a tuple of proper tokens
        field : str
            field the word falls under
        smoother : str (optional)
            one of {None, 'offset', 'log'}
        offset : int
            offset to add to count only needed for smoothers 'log' and 'offset'

        Returns
        -------
        float
        """
        if isinstance(term, str):
            words = self.pre_process(term, self.preprocessors[field])
        else:
            words = term

        smoothers = {
            None: self._smooth_none,
            'offset': self._smooth_offset,
            'log': self._smooth_log
        }
        if smoother not in smoothers:
            raise KeyError(
                'Chosen smother `{}` is not one of {}'.format(
                    smoother, smoothers.keys())
            )
        counter = self.field_freq[field]

        s_fun = smoothers[smoother]
        max_occ = self._most_common[field]
        id_potentials = {
            word: s_fun(counter[word], offset, max_occ)
            for word in words
        }

        total_weight = sum(id_potentials.values())
        return {w: idp / total_weight for w, idp in id_potentials.items()}

    def alias_graph(self, threshold=0.7, verbose=True, weights=None, **kwargs) -> DiGraph:
        """
        This function creates the directed network connecting an instance to an other
        through a directed edge if the the target instance has a similarity score above
        the threshold.

        Parameters
        ----------
        weights
        threshold : float
            between 0 and 1
        verbose : whether to show the progressbar
        kwargs :
            arguments to pass to the score function (i.e. offset, smoother)

        Returns
        -------
        DiGraph

        """
        if not verbose:
            def t(x):
                return x
        else:
            t = tqdm

        edges = []
        fields = self.records.keys()
        a_field = list(fields)[0]
        for idx in t(self.records[a_field]):
            targets = self.find(idx, threshold=threshold, weights=weights, **kwargs)
            new_edges = [(idx, x.index, {'score': x.score}) for x in targets]
            edges.extend(new_edges)

        alias_graph = DiGraph()
        alias_graph.add_edges_from(edges)
        return alias_graph

    def disambiguated_sets(self, threshold=0.7, verbose=True, weights=None, **kwargs):
        from networkx import strongly_connected_components
        alias_graph = self.alias_graph(threshold, verbose, weights, **kwargs)
        strong = strongly_connected_components(alias_graph)
        return list(strong)

    @staticmethod
    def pre_process(base_name, functions: list):
        """apply every function consecutively to base_name"""
        norm_name = base_name
        for f in functions:
            norm_name = f(norm_name)
        return set(norm_name)

    @property
    def fields(self):
        return self.field_freq.keys()

    @staticmethod
    def _smooth_none(occurrences, *args):
        return 1 / max(occurrences, 1)

    @staticmethod
    def _smooth_offset(occurrences, offset, *args):
        return 1 / max(occurrences + offset, 1)

    @staticmethod
    def _smooth_log(occ, offset, max_occ):
        max_offset = max(max_occ + offset + 1, 1)
        word_offset = max(occ + offset, 1)
        return log(max_offset / word_offset)
